### PR TITLE
fix network_cli timeout

### PR DESCRIPTION
**Pull request by f-bor**
##### SUMMARY
PR #47345 and #47004 are causing a random timeout to occur. By removing either, the timeout disappears. I suppose this is a problem between signals and paramiko's critical sections/conditions used in recv_ready, or maybe time.sleep. However, there is no need anymore to use signals to manage the buffer_read timeout since #47345 introduce such management.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
network_cli

##### ANSIBLE VERSION

```paste below
ansible 2.8.0.dev0 (timeout_fix 79989cbda5) last updated 2018/10/27 09:42:05 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/fbor/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/fbor/ansible/lib/ansible
  executable location = /home/fbor/ansible/bin/ansible
  python version = 2.7.13 (default, Sep 26 2018, 18:42:22) [GCC 6.3.0 20170516]
```

##### ADDITIONAL INFORMATION
To reproduce the issue, just setup an integration environnement for vyos and run:
```
ansible-test network-integration -vvvv --inventory inventory.networking vyos_.*
```

EDIT: After digging a little bit more, the timeout is caused by a deadlock. It seems that when signaled, python's execution is interrupted and does not resume in certain cases. Then, Paramiko's lock is staying in an acquired state and never get released. Here is a trivial way to reproduce the receive loop and the deadlock:

```
import signal
import time
import threading
import math

class DeadLock:

    def __init__(self):
        self.lock = threading.Lock()

    def sighandler(self, signum, frame):
        print('signaled')
        raise Exception

    def compute(self):
        print('locking')
        self.lock.acquire()
        print('locked')
        try:
            return math.acos(1)
        finally:
            print('unlocking')
            self.lock.release()
            print('unlocked')
            return -1


    def run(self):
        while True:
            signal.signal(signal.SIGALRM, self.sighandler)
            signal.setitimer(signal.ITIMER_REAL, 0.1)
            try:
                while True:
                    self.compute()
                    sleep(0.001)
            except:
                pass

dl = DeadLock()
dl.run()
```

It always gives this after a while:
> locking
> locked
> unlocking
> signaled
> locking

### CODE REVIEW

File: lib/ansible/plugins/connection/network_cli.py [hunk 1]
```diff
@@ -380,9 +380,6 @@ def receive(self, command=None, prompts=None, answer=None, newline=True, prompt_
         command_prompt_matched = False
         matched_prompt_window = window_count = 0
 
-        command_timeout = self.get_option('persistent_command_timeout')
```
Comment (ganeshrn) [inline lib/ansible/plugins/connection/network_cli.py:373]: Removing the signal from here won't entirely solve the issue as `persistent_command_timeout` timer signal is also triggered from `ansible-connection` which is common for all the connection plugins using the persistent framework (netconf/network_cli/httpapi). This mix of signal and sleep in `network_cli` is resulting in non-deterministic behavior, based on this I think we need to revert PR https://github.com/ansible/ansible/pull/47345 as the signal based timeout handling is in place since 2.3 and is working fine.

Comment (f-bor) [reply lib/ansible/plugins/connection/network_cli.py:373]: I have to disagree. You can remove the sleep call from the poc I gave, it will just deadlock faster. With my limited understanding of ansible's architecture and python, using signal from ansible-connection for persistent_command_timeout makes sense because it's a kind of last resort measure, it does not happens under normal conditions or during regular network activity and it has a very long delay. So the likelihood to trigger that bug from there is about 0. With a delay of 100ms and just in the middle of regular network activity, it's just a matter of time or network conditions to let a lock acquired in paramiko. However the probability should be small (because it needs to already had a prompt, because paramiko will most of the time be seating in wait at the signal trigger, and because wait is releasing the lock before sleeping). With the recv_ready calls every 1ms, it is a certainty. But in the end, using something such violent as signaling to just manage a small network timeout under normal circonstances does not sound like good design (unless you're writing some kind of lowlevel library) so such problems should be expected.

Comment (ganeshrn) [reply lib/ansible/plugins/connection/network_cli.py:373]: > it does not happen under normal conditions or during regular network activity and it has 
  a very long delay

`persistent_command_timeout` default timeout value triggered from `ansible-connection` is 10 secs, not sure how you concluded it is a 'very long delay'. 

>  With the recv_ready calls every 1ms, it is a certainty. But in the end, using something such violent as signaling to just manage a small network timeout under normal circonstances does not sound like good design 

One of the resaon to add a signal to check if comeplete response is reecevied or not is because the signal handler for `persistent_command_timeout`  trigerred from `ansible-connection` was not invoked in case of sleep().

Comment (f-bor) [reply lib/ansible/plugins/connection/network_cli.py:373]: By very long delay, I was meaning relative to the other signal delay of 100ms. But once again, since persistent_command_timeout triggers the end of program, I dont see the issue here.

Also, I'm not sure to understand the sleep issue. Before #47345, you were signaling around paramiko's recv, which will be sleeping nearly all the time when signal handler is called. I have read signal may not interrupts a sleep call and may triggers only when it has finished. But who cares when sleep just lasting few milliseconds ? Is this about this behavior or am i missing something here ?

In summary, in my opinion, the issue is more related to finally clauses and using signal with exceptions. When paramiko is already in the middle of a finally clause, cleaning everything up (like going to release a lock), when at that exact time the signal handler is called, since it raises an exception the finally clause can't finish her work and leaves in a dirty state (like with a lock acquired). Don't you agree that's the issue's root ?

Comment (ganeshrn) [reply lib/ansible/plugins/connection/network_cli.py:373]: It seems network integration test is failing because of PR #47345, hence reverting the PR

Comment (ganeshrn) [reply lib/ansible/plugins/connection/network_cli.py:373]: The 0.001 sleep doesn't seem to to be sufficient for recv_ready() and after couple of runs results in an deadlock. 
Looking at performance improvement with this check it seems it not that significant at scale unless I am missing something. Considering 2sec/per command (optimistic estimation) for 1000 command will take around ~33.33 mins to complete playbook run. With this check added considering the best improvement estimate will be 1.95 sec/per command that is ~32.5 mins.

The sleep in recv() doesn't result in deadlock, even if the `persistent_command_timeout` signal handler is missed it is bailed out from while loop by `socket.timeout` execption handling.

Comment (f-bor) [reply lib/ansible/plugins/connection/network_cli.py:373]: > The 0.001 sleep doesn't seem to to be sufficient for recv_ready() and after couple of runs results in an deadlock.

This has little to do with sleep length. Increasing sleep time just lowering probability that a finally clause get destroyed by the exception raised in signal handler, but just because you will run less finally clauses.
 
> Looking at performance improvement with this check it seems it not that significant at scale unless I am missing something. Considering 2sec/per command (optimistic estimation) for 1000 command will take around ~33.33 mins to complete playbook run. With this check added considering the best improvement estimate will be 1.95 sec/per command that is ~32.5 mins.

2 seconds per command ? On my cheap network hardware, a command typically ran at 100ms during my tests (including ~55ms of ssh time without #47345). When you declare a vlan on it, it sets all the switch interfaces on dynamically registered through GVRP, so you have to exclude the vlan everywhere you don't want it to have well defined vlans. With 48 ports switchs and some vlans, it generates a lot of commands. Gaining 25% is not negligeable on that part.
 
> The sleep in recv() doesn't result in deadlock, even if the `persistent_command_timeout` signal handler is missed it is bailed out from while loop by `socket.timeout` execption handling.

Recv also has a finally clause that is also releasing a lock just like recv_ready. And therefore it can cause a deadlock like recv_ready with signals raising exceptions. It's just a lot less likely to happen because there is only one recv call (and one finally clause run) for every signal trigger.
